### ISSUE TITLE
Efficiently compute reduced potential at a list of states

### ISSUE THREAD

**Issue by andrrizzi** [choderalab/openmmtools#321]
With exact treatment of PME electrostatics, computing the energy matrix required for MBAR is very expensive. We can compute the optimization discussed in choderalab/yank#678 and choderalab/sams#9: i.e., split the alchemical forces into different groups and compute the energy of only the groups that are different. Here's how I plan to add the feature:

- Add option `split_force_groups` in `AbsoluteAlchemicalFactory.__init__()` to split forces by alchemical variable.
- Require `ComposableStates` to implement a method `_find_force_groups_to_update(context, context_thermodynamic_state)` that figures out which force groups need to be computed to obtain the difference of potential energy.
- Add a classmethod `ThermodynamicState.reduced_potential_at_states(context, thermodynamic_state_list)` that uses the method above to efficiently compute the reduced potential at all the given thermodynamic states.

**Comment by Lnaden** [comment]
You could make `ThermodynamicState` have a lookup table basically tracking which force is in which group and just let `ComposableStates` access some `ThermodynamicState` API instead of computing it on-the-fly all the time

**Comment by jchodera** [comment]
Maybe we should post on the OpenMM issue tracker too? OpenMM already does some things under the hood to speed up energy and force computation for integrators when recomputation isn't needed. What if we were able to do the same thing for individual forces? Right now, openmm invalidates all forces when context parameters change instead of just invalidating the forces that have changed. It might be simple to just put this in at the openmm level.

**Comment by jchodera** [comment]
The above proposal does look nice, though!

**Comment by jchodera** [comment]
Whoops, pressed the wrong button! Sorry!

**Comment by andrrizzi** [comment]
Good idea! Usually cycling through the forces of a System is super quick, but I can implement something like that if it makes sense. I'll check the timings.

**Comment by andrrizzi** [comment]
> Maybe we should post on the OpenMM issue tracker too?

I'll post also there, thanks! I'd still implement this though as we need it ASAP. The function can become a simple shortcut once the OpenMM features will be added.

**Comment by jchodera** [comment]
> I'll post also there, thanks! I'd still implement this though as we need it ASAP. The function can become a simple shortcut once the OpenMM features will be added.

Sounds good!
